src/config/plugin.ts

```diff
@@ -284,26 +355,87 @@ export class Plugin implements IPlugin {
   }
 
   private async getCommandIDs(): Promise<string[]> {
-    const commandsDir = await this.getCommandsDir()
-    if (!commandsDir) return []
-
     const marker = Performance.mark(OCLIF_MARKER_OWNER, `plugin.getCommandIDs#${this.name}`, {plugin: this.name})
-    this._debug(`loading IDs from ${commandsDir}`)
-    const files = await globby(GLOB_PATTERNS, {cwd: commandsDir})
-    const ids = processCommandIds(files)
+
+    let ids: string[]
+    switch (this.commandDiscoveryOpts?.strategy) {
+      case 'explicit': {
+        ids = (await this.getCommandIdsFromTarget()) ?? []
+        break
+      }
+
+      case 'pattern': {
+        ids = await this.getCommandIdsFromPattern()
+        break
+      }
+
+      case 'single': {
+        ids = (await this.getCommandIdsFromTarget()) ?? []
+        break
+      }
+
+      default: {
+        ids = []
+      }
+    }
+
     this._debug('found commands', ids)
     marker?.addDetails({count: ids.length})
     marker?.stop()
     return ids
   }
 
+  private async getCommandIdsFromPattern(): Promise<string[]> {
+    const commandsDir = await this.getCommandsDir()
+    if (!commandsDir) return []
+
+    this._debug(`loading IDs from ${commandsDir}`)
+    const files = await globby(this.commandDiscoveryOpts?.globPatterns ?? GLOB_PATTERNS, {cwd: commandsDir})
+    return processCommandIds(files)
+  }
+
+  private async getCommandIdsFromTarget(): Promise<string[] | undefined> {
+    const commandsFromExport = await this.loadCommandsFromTarget()
+    if (commandsFromExport) {
+      return (
+        Object.entries(commandsFromExport)
+          .map(([id, cmd]) => {
+            if (!ensureCommandClass(cmd)) return
+            return id
+          })
+          // eslint-disable-next-line unicorn/prefer-native-coercion-functions
+          .filter((f): f is string => Boolean(f))
+      )
+    }
+  }
```

Comment:
if you filter first you don't have to deal with undefineds.  You can also modify the method to always return an array since the 2 consumers are doing `?? []` after it anyway.

```suggestion
  private async getCommandIdsFromTarget(): Promise<string[]> {
    return Object.entries((await this.loadCommandsFromTarget()) ?? [])
      .filter(([, cmd]) => ensureCommandClass(cmd))
      .map(([id]) => id)
  }
```